### PR TITLE
Show best trial number and value in progress bar

### LINKED ISSUE #3661: Add best score reporting into progress bar
### Motivation

It becomes quite inconvenient to scroll through hundreds of optuna run logs in jupyter notebook to find what is the current best score(s). At the same time, progress bar always stays on  top of the output cell, and it would be easy to track current best score via progress bar description line. Also, it is already implemented in LightGBM integration: https://github.com/optuna/optuna/blob/2d6c15ee68065132070862b66de14894f35294ad/optuna/integration/_lightgbm_tuner/optimize.py#L205

### Description

Add the current best score(s) reporting to the progress bar's description or postfix string, when the progress bar is used.

### Alternatives (optional)

1. Look at the run logs, provided by optuna.
2. Save optimization runs info into some persistent storage - SQLite, Redis, RDBMS and track best scores via optuna-dashboard
3. Track scores by using MLFlow/WandB integration

### Additional context (optional)

_No response_

### PR DESCRIPTION
**Pull request by nzw0301**
## Motivation

Resolve #3661 and takeover #4093, which introduces a new progress bar stat info: best trial value.

## Description of the changes

When the objective is single, show best trial number and value in tqdm's progress bar.

Difference from the original PR
1. Apply reviewers' suggestions on the original PR
2. Add tests

### CODE REVIEW

**Comment by nzw0301** [comment]
I cannot reproduce the error on my local machine...

**Comment by contramundum53** [comment]
@not522 Could you review this PR?

File: optuna/progress_bar.py [hunk 1]
```diff
@@ -75,15 +77,21 @@ def _init_valid(self) -> None:
...
 
         Args:
             elapsed_seconds:
                 The time past since :func:`~optuna.study.Study.optimize` started.
+            study:
+                The current study object.
         """
 
         if self._is_valid:
+            if not study._is_multi_objective() and study.get_trials(states=[TrialState.COMPLETE]):
```
Comment (not522) [inline optuna/progress_bar.py:91]: This line seems to be slowing down the process.

**Comment by not522** [comment]
This PR seems to be several times slower than master.

File: optuna/progress_bar.py (hunk 1 above)
Comment (not522) [reply optuna/progress_bar.py:91]: This PR seems to be several times slower than master.

**Comment by not522** [comment]
I rerun the CI, and it passed.

File: optuna/progress_bar.py (hunk 1 above)
Comment (nzw0301) [reply optuna/progress_bar.py]: Thank you for your comments! Indeed, the original code was severely slower than the master branch. I've updated this part.

File: optuna/progress_bar.py [hunk 2]
```diff
@@ -75,15 +76,27 @@ def _init_valid(self) -> None:
...
+            study:
+                The current study object.
         """
 
         if self._is_valid:
+            if not study._is_multi_objective():
+                # Not updating the progress bar when there are no complete trial.
+                try:
+                    msg = f"Best trial: {study.best_trial.number}. \
+                            Best value: {study.best_value:.6g}"
```
Comment (not522) [inline optuna/progress_bar.py:94]: It seems unexpected spaces.

```suggestion
                    msg = (
                        f"Best trial: {study.best_trial.number}. "
                        f"Best value: {study.best_value:.6g}"
                    )
```

**Review by not522 — COMMENTED** [review]
Thank you for your update. It's almost LGTM. Please check my suggestion.

File: optuna/progress_bar.py (hunk 2 above)
Comment (nzw0301) [reply optuna/progress_bar.py]: Thank you for your careful check! I've accepted your suggestion.

**Review by not522 — APPROVED** [review]
LGTM!

**Review by contramundum53 — APPROVED** [review]
LGTM!

**Pull request merged**